File: java/kata/src/test/java/br/com/ari/algo/structures/StackTest.java
package br.com.ari.algo.structures;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class StackTest {

    @Test
    void testStack() {
        Stack<Integer> stack = new Stack<>();
        stack.push(10);
        stack.push(20);
        stack.push(30);
        stack.push(40);
        stack.push(50);

        assertThat(stack.size()).isEqualTo(5);
        assertThat(stack.peek()).isEqualTo(50);

        assertThat(stack.pop()).isEqualTo(50);

        assertThat(stack.size()).isEqualTo(4);
        assertThat(stack.peek()).isEqualTo(40);

        int[] expectedValues = new int[] { 40, 30, 20, 10 };
        for (int expected : expectedValues) {
            assertThat(stack.pop()).isEqualTo(expected);
        }

        assertThat(stack.size()).isZero();
    }
    
}

File: go/kata/structures/stack_test.go
package structures

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestStack(t *testing.T) {
	stack := NewStack[int]()
	assert.Equal(t, 0, stack.Peek())
	assert.Equal(t, 0, stack.Pop())

	stack.Push(1)
	stack.Push(2)
	stack.Push(3)
	stack.Push(4)
	stack.Push(5)

	assert.Equal(t, 5, stack.Peek())
	assert.Equal(t, 5, stack.Peek())
	assert.Equal(t, 5, stack.Pop())
	assert.Equal(t, 4, stack.Pop())
	assert.Equal(t, 3, stack.Pop())
	assert.Equal(t, 2, stack.Pop())
	assert.Equal(t, 1, stack.Pop())
	assert.Equal(t, 0, stack.Pop())
	assert.Equal(t, 0, stack.Pop())

	stack.Push(10)
	stack.Push(20)
	stack.Push(30)
	stack.Push(40)
	stack.Push(50)

	assert.Equal(t, 50, stack.Pop())
	assert.Equal(t, 40, stack.Pop())
	assert.Equal(t, 30, stack.Pop())

	stack.Push(100)

	assert.Equal(t, 100, stack.Pop())
	assert.Equal(t, 20, stack.Pop())
	assert.Equal(t, 10, stack.Pop())
	assert.Equal(t, 0, stack.Pop())
}
